notify/impl.go

```diff
@@ -875,6 +875,9 @@ func (n *Wechat) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
 		parameters := url.Values{}
 		parameters.Add("corpsecret", tmpl(string(n.conf.APISecret)))
 		parameters.Add("corpid", tmpl(string(n.conf.CorpID)))
+		if err != nil {
```

Comment:
I am confused here. Isn't `err` last assigned in line 868? Am I missing something?

Reply:
This is indeed confusing (I've had the same moment myself) but it's because the `tmpl` function holds a reference to `err` (see L863). Thus `err` needs to be checked after the `tmpl` function has been called. If you look at the other notifiers, they use the same pattern.

https://github.com/prometheus/alertmanager/blob/dac673a6aa183a2a0787e63e8cfa1701965975b6/notify/impl.go#L688-L698

That being said, we could have a dedicated `error` variable (eg `tmplErr`) to make it easier to understand.

Reply:
BTW I'll take a pass at the other notifiers because there might be other places where error checking isn't done properly.

Reply:
Thanks for clarifying. This is a very confusing implicit coding style to me.

Instead of `tmplText` return a function with return value of `(s string)` why not have it return a function that has its own error as return values like `(s string, error)`?

In case I am not missing something and enough people agree I can create a follow up PR. What do you think @simonpasquier @stuartnelson3?

Reply:
@mxinden I definitely agree. I've tested an approach which is a bit different from what you proposed in https://github.com/prometheus/alertmanager/pull/1436/commits/2c06ad1a0c1b8ef4390253fff008dea71f10cf2c.